Added more features to `CORS`

girardinsamuel opened the pull request:
This PR improve Masonite CORS operations protection capabilities (Fix #476)

- [x] Add a real `cors` feature as any other Masonite feature
- [x] Use this feature in the existing `CorsMiddleware`
- [x] Add unit tests
- [x] Add docs: you can find a preview here https://docs.masoniteproject.com/v/development/security/cors
- [x] Add config to new projects: https://github.com/MasoniteFramework/cookie-cutter/pull/46

girardinsamuel committed:
add cors service

girardinsamuel committed:
use cors service in middleware

girardinsamuel committed:
start adding cors unit tests

josephmancuso commented:
Agreed. I like where this is heading

girardinsamuel committed:
update tests

girardinsamuel committed:
add ability to respond to options request and to return method not allowed response

girardinsamuel committed:
add ability to answer to HEAD requests

girardinsamuel committed:
remove unused import

girardinsamuel committed:
fix head

girardinsamuel committed:
fix building response

girardinsamuel committed:
fix withHeaders withCookies and add dumps features to TestCase

girardinsamuel committed:
fix

girardinsamuel committed:
improve code

girardinsamuel committed:
remove tests to avoid cluttering test outputs

girardinsamuel committed:
update docstrings

girardinsamuel committed:
remove dump statement

girardinsamuel committed:
Merge branch 'feat/fix-testcase' into feat/476

girardinsamuel committed:
add more tests

josephmancuso commented on src/masonite/routes/HTTPRoute.py:
```diff
@@ -30,7 +29,12 @@ def __init__(
+
         self.request_method = [x.lower() for x in request_method]
+        # add HEAD method if GET
+        if "get" in self.request_method and "head" not in self.request_method:
```
This part is confusing to me. Is there another way to do this? Seems weird automatically adding head to every GET route

girardinsamuel replied:
That 's how Laravel is handling it and that's quite mandatory because each GET route should be able to be accessed through GET !

josephmancuso commented on src/masonite/routes/Route.py:
```diff
@@ -24,7 +24,7 @@ def get(self, url, controller, module_location=None, **options):
             url,
             controller,
-            request_method=["get"],
+            request_method=["get", "head"],
```
Not sure i like doing this. This feature seems to tightly coupled to other features

girardinsamuel replied:
Above item and this one should be discussed in the PR https://github.com/MasoniteFramework/masonite/pull/578.

The HEAD thing is a prerequisite to this PR and is implemented in #578.

josephmancuso commented on tests/features/cors/test_cors.py:
```diff
@@ -0,0 +1,143 @@
+        self.application.make("middleware").add([CorsMiddleware])
+        self.cors = self.application.make("cors")
+        self.default_options = {
+            "allowed_origins": ["http://localhost"],
```
Good to whichever origin matches an element in this list, thats the host that is returned in the origin header right?

girardinsamuel replied:
I did not understand the question

josephmancuso replied:
I have no idea what I was trying to ask here lol

josephmancuso commented on tests/integrations/config/cors.py:
```diff
@@ -0,0 +1,15 @@
+
+ALLOWED_METHODS = ["*"]
+
+ALLOWED_ORIGINS = ["*"]
```
I think we need to have these configs more similar to our other configs. 

Something like:

```
CORS = {
   "paths": ["api/*", "sanctum/csrf-cookie", "server1.com"],
   "allowed_methods": ["*"],
   ....
   
}
```

girardinsamuel replied:
The issue is with this is that it's already namespaced with cors through the config file name...so accessing it later will result in doing `config("cors.cors.paths")`....

josephmancuso replied:
The rest of the PR looks good but still not sold on this.

> The issue is with this is that it's already namespaced with cors through the config file name...so accessing it later will result in doing `config("cors.cors.paths")`....

You have a valid point here. Honestly it looks better than this config file though. This looks very M3 to me.

Just throwing some idea out there but do we have like a `security.py` file so at least it would be `config("security.cors.paths")`? And we can put other security settings in there either now or later

girardinsamuel replied:
Hmm we don't have it yet. But we could add it, `cors` was already new so we could change it to `security` why not, i like it

josephmancuso commented:
@girardinsamuel if you can, can you resolve all these conflicts? I merged in the PR for the options and HEAD

girardinsamuel committed:
Merge branch '4.0' into feat/476

girardinsamuel committed:
fix lint

girardinsamuel committed:
fix tests

girardinsamuel committed:
update default cors optiongs

girardinsamuel commented:
It's done. I just need to update documentation.
And maybe the way cors options are defined as you said in your review.

girardinsamuel committed:
update default cors options

josephmancuso commented:
Just not sold on the config file

girardinsamuel committed:
move config to security

girardinsamuel commented:
> Just not sold on the config file

I updated to what you proposed. cookie-cutter and docs are updated too
👉 https://docs.masoniteproject.com/v/development/security/cors#configuration you're right it's cleaner

josephmancuso reviewed (approved):
Code review looks good. Will need to test this manually.

One thing I want to be sure of is what happens when there is no security.py file

girardinsamuel committed:
Merge remote-tracking branch 'upstream/4.0' into feat/476

girardinsamuel committed:
add default options for old projects

girardinsamuel commented:
> Code review looks good. Will need to test this manually.
> 
> One thing I want to be sure of is what happens when there is no security.py file

You're right. I have added some default options for old projects ✅. In Masonite 5 we will be able to remove the lines I added. And maybe create a SecurityProvider for registering `cors` feature.
